File: notifications-polyglot/typescript/src/notification/notification-status.service.ts
import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { NotificationStatus } from './notification-status.entity';
import { NotificationMetrics } from './notification-metrics.dto';

@Injectable()
export class NotificationStatusService {
    constructor(
        @InjectRepository(NotificationStatus)
        private statusRepository: Repository<NotificationStatus>,
    ) { }

    async createInitialStatus(notificationId: string, channels: string[]): Promise<void> {
        const channelStatuses: Record<string, string> = {};
        channels.forEach((channel) => {
            channelStatuses[channel] = 'PENDING';
        });

        const status = this.statusRepository.create({
            notification_id: notificationId,
            status: 'ENQUEUED',
            channel_statuses: channelStatuses,
        });

        await this.statusRepository.save(status);
    }

    async getStatus(notificationId: string): Promise<NotificationStatus> {
        return this.statusRepository.findOneBy({ notification_id: notificationId });
    }

    async updateChannelStatus(
        notificationId: string,
        channel: string,
        newStatus: string,
    ): Promise<void> {
        const status = await this.statusRepository.findOneBy({ notification_id: notificationId });
        if (!status) {
            return;
        }

        status.channel_statuses[channel] = newStatus;

        // Aggregate overall status
        let allCompleted = true;
        let anyFailed = false;

        Object.values(status.channel_statuses).forEach((s) => {
            if (s !== 'COMPLETED') {
                allCompleted = false;
            }
            if (s === 'FAILED') {
                anyFailed = true;
            }
        });

        if (allCompleted) {
            status.status = 'COMPLETED';
        } else if (anyFailed) {
            status.status = 'PARTIAL_FAILURE';
        } else {
            status.status = 'PROCESSING';
        }

        await this.statusRepository.save(status);
    }

    async getMetrics(): Promise<NotificationMetrics> {
        const results = await this.statusRepository
            .createQueryBuilder('status')
            .select('status.status', 'status')
            .addSelect('COUNT(*)', 'count')
            .groupBy('status.status')
            .getRawMany();

        const counts: Record<string, number> = {};
        let total = 0;
        let success = 0;
        let failed = 0;

        results.forEach((r) => {
            const count = parseInt(r.count, 10);
            counts[r.status] = count;
            total += count;
            if (r.status === 'COMPLETED') {
                success += count;
            } else if (r.status === 'FAILED') {
                failed += count;
            }
        });

        return {
            total_sent: total,
            total_success: success,
            total_failed: failed,
            by_status: counts,
        };
    }
}

File: notifications-polyglot/golang/service/notification_status_service.go
package service

import (
	"strings"
	"time"

	"gorm.io/gorm"
)

type NotificationStatusService struct {
	db *gorm.DB
}

func NewNotificationStatusService(db *gorm.DB) *NotificationStatusService {
	return &NotificationStatusService{db: db}
}

func (s *NotificationStatusService) CreateInitialStatus(notificationID string, channels []string) error {
	channelStatuses := make(map[string]string)
	for _, channel := range channels {
		channelStatuses[channel] = "PENDING"
	}

	status := &NotificationStatus{
		NotificationID:  notificationID,
		Status:          "ENQUEUED",
		ChannelStatuses: channelStatuses,
		CreatedAt:       time.Now(),
		UpdatedAt:       time.Now(),
	}

	return s.db.Create(status).Error
}

func (s *NotificationStatusService) GetStatus(notificationID string) (*NotificationStatus, error) {
	var status NotificationStatus
	err := s.db.First(&status, "notification_id = ?", notificationID).Error
	if err != nil {
		return nil, err
	}
	return &status, nil
}

func (s *NotificationStatusService) UpdateChannelStatus(notificationID, channel, newStatus string) error {
	var status NotificationStatus
	if err := s.db.First(&status, "notification_id = ?", notificationID).Error; err != nil {
		return err
	}

	status.ChannelStatuses[channel] = newStatus

	// Aggregate overall status
	allCompleted := true
	anyFailed := false

	for _, s := range status.ChannelStatuses {
		if !strings.EqualFold(s, "COMPLETED") {
			allCompleted = false
		}
		if strings.EqualFold(s, "FAILED") {
			anyFailed = true
		}
	}

	if allCompleted {
		status.Status = "COMPLETED"
	} else if anyFailed {
		status.Status = "PARTIAL_FAILURE"
	} else {
		status.Status = "PROCESSING"
	}

	status.UpdatedAt = time.Now()
	return s.db.Save(&status).Error
}

func (s *NotificationStatusService) GetMetrics() (*NotificationMetrics, error) {
	var results []struct {
		Status string
		Count  int64
	}

	err := s.db.Model(&NotificationStatus{}).
		Select("status, count(*) as count").
		Group("status").
		Scan(&results).Error

	if err != nil {
		return nil, err
	}

	counts := make(map[string]int64)
	var total, success, failed int64

	for _, r := range results {
		counts[r.Status] = r.Count
		total += r.Count
		if strings.EqualFold(r.Status, "COMPLETED") {
			success += r.Count
		} else if strings.EqualFold(r.Status, "FAILED") {
			failed += r.Count
		}
	}

	return &NotificationMetrics{
		TotalSent:    total,
		TotalSuccess: success,
		TotalFailed:  failed,
		ByStatus:     counts,
	}, nil
}
